apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts

```diff
@@ -0,0 +1,323 @@
+
+            resolve();
+          } catch (error) {
+            reject(error);
+          }
+        }
+      });
+    });
+  }
+
+  /**
+   * This is called once for each local project from Rush.json.
+   * @param project             The local project that we will create symlinks for
+   * @param commonRootPackage   The common/temp/package.json package
+   * @param commonPackageLookup A dictionary for finding packages under common/temp/node_modules
+   * @param rushConfiguration   The rush.json file contents
+   * @param rushLinkJson        The common/temp/rush-link.json output file
+   * @param options             Command line options for "rush link"
+   */
+  private _linkProject(
+    project: RushConfigurationProject,
+    commonRootPackage: NpmPackage,
+    commonPackageLookup: PackageLookup,
+    rushLinkJson: IRushLinkJson): void {
+
+    let commonProjectPackage: NpmPackage | undefined =
+      commonRootPackage.getChildByName(project.tempProjectName) as NpmPackage;
+    if (!commonProjectPackage) {
+      // Normally we would expect the temp project to have been installed into the common\node_modules
+      // folder.  However, if it was recently added, "rush install" doesn't technically require
+      // this, as long as its dependencies can be found at the root of the NPM shrinkwrap file.
+      // This avoids the need to run "rush generate" unnecessarily.
+
+      // Example: "project1"
+      const unscopedTempProjectName: string = Utilities.parseScopedPackageName(project.tempProjectName).name;
+
+      // Example: "C:\MyRepo\common\temp\projects\project1
+      const extractedFolder: string = path.join(this._rushConfiguration.commonTempFolder,
+        RushConstants.rushTempProjectsFolderName, unscopedTempProjectName);
+
+      // Example: "C:\MyRepo\common\temp\projects\project1.tgz"
+      const tarballFile: string = path.join(this._rushConfiguration.commonTempFolder,
+        RushConstants.rushTempProjectsFolderName, unscopedTempProjectName + '.tgz');
+
+      // Example: "C:\MyRepo\common\temp\projects\project1\package.json"
+      const packageJsonFilename: string = path.join(extractedFolder, 'package', 'package.json');
+
+      Utilities.createFolderWithRetry(extractedFolder);
+      tar.extract({
+        cwd: extractedFolder,
+        file: tarballFile,
+        sync: true
+      });
+
+      // Example: "C:\MyRepo\common\temp\node_modules\@rush-temp\project1"
+      const installFolderName: string = path.join(this._rushConfiguration.commonTempFolder,
+        RushConstants.nodeModulesFolderName, RushConstants.rushTempNpmScope, unscopedTempProjectName);
+
+      commonProjectPackage = NpmPackage.createVirtualTempPackage(packageJsonFilename, installFolderName);
+
+      // remove the extracted tarball contents
+      fsx.removeSync(packageJsonFilename);
+      fsx.removeSync(extractedFolder);
+
+      commonRootPackage.addChild(commonProjectPackage);
+    }
```

Comment:
extract this?

Reply:
Perhaps I will do this in the next PR?

---
In reply to: [152137695](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152137695) [](ancestors = 152137695)